src/System.Windows.Forms.Design/src/System/Windows/Forms/Design/FlowLayoutPanelDesigner .cs

```diff
@@ -578,138 +578,146 @@ private void ReorderControls(DragEventArgs de)
+                }
+            }
+
+            if (_insertionIndex == s_invalidIndex)
             {
-                continue;
+                // Either _insertionIndex was _childInfo.Length (inserting past the end) or
+                // _insertionIndex was _childInfo.Length - 1 and the control at that index was also
+                // a part of the dragCollection. In either case, the new index is equal to the count
+                // of existing controls in the ControlCollection. Helps to draw this out.
+                _insertionIndex = children.Count;
             }
 
-            children.Add(_dragControls[i]);
-            children.SetChildIndex(_dragControls[i], _insertionIndex);
+            children.Add(_primaryDragControl);
+            children.SetChildIndex(_primaryDragControl, _insertionIndex);
             ++_insertionIndex;
 
-            SelectionService.SetSelectedComponents(new IComponent[] { _dragControls[i] }, SelectionTypes.Add);
-        }
+            //Set the Selection ..
+            SelectionService.SetSelectedComponents(new IComponent[] { _primaryDragControl }, SelectionTypes.Primary | SelectionTypes.Replace);
 
-        if (controlsProperty is not null)
-        {
-            RaiseComponentChanging(controlsProperty);
-        }
+            // Note _dragControls are in opposite order than what FLP uses,
+            // so add from the end.
+            for (var i = _dragControls.Count - 1; i >= 0; i--)
+            {
+                if (_primaryDragControl.Equals(_dragControls[i]))
+                {
+                    continue;
+                }
 
-        // If we did a Copy, then restore the old controls to make sure we set state correctly.
-        if (originalControls is not null)
-        {
-            for (var i = 0; i < originalControls.Count; i++)
+                children.Add(_dragControls[i]);
+                children.SetChildIndex(_dragControls[i], _insertionIndex);
+                ++_insertionIndex;
+
+                SelectionService.SetSelectedComponents(new IComponent[] { _dragControls[i] }, SelectionTypes.Add);
+            }
+
+            if (controlsProperty is not null)
             {
-                _dragControls[i] = (Control)originalControls[i];
+                RaiseComponentChanging(controlsProperty);
             }
-        }
 
-        base.OnDragComplete(de);
+            // If we did a Copy, then restore the old controls to make sure we set state correctly.
+            if (originalControls is not null)
+            {
+                for (var i = 0; i < originalControls.Count; i++)
+                {
+                    _dragControls[i] = (Control)originalControls[i];
+                }
+            }
+
+            base.OnDragComplete(de);
 
-        designerTransaction.Commit();
+            designerTransaction?.Commit();
+            designerTransaction = null;
+        }
+        finally
+        {
+            designerTransaction?.Cancel();
+        }
```

Comment:
Maybe it's better to wrap `designerTransaction?.Cancel();` in `catch` block.
```suggestion
        catch
        {
            designerTransaction?.Cancel();
        }
```